src/Plugin/Field/FieldFormatter/IslandoraImageFormatter.php

```diff
@@ -103,39 +116,119 @@ public static function create(ContainerInterface $container, array $configuratio
+    $alt_text_options = [
+      'local' => $this->t('Local'),
+      'original_file_fallback' => $this->t('Local, with fallback to Original File'),
+      'original_file' => $this->t('Original File'),
+    ];
+    $element['image_alt_text'] = [
+      '#title' => $this->t('Alt text source'),
+      '#type' => 'select',
+      '#default_value' => $this->getSetting('image_alt_text'),
+      '#empty_option' => $this->t('None'),
+      '#options' => $alt_text_options,
+    ];
+    return $element;
+  }
+
   /**
    * {@inheritdoc}
    */
   public function viewElements(FieldItemListInterface $items, $langcode) {
     $elements = parent::viewElements($items, $langcode);
 
     $image_link_setting = $this->getSetting('image_link');
-    // Check if the formatter involves a link.
-    if ($image_link_setting != 'content') {
+    $alt_text_setting = $this->getsetting('image_alt_text');
+    // Check if we can leave the image as-is:
+    if ($image_link_setting != 'content' && $alt_text_setting == 'local') {
       return $elements;
     }
-
     $entity = $items->getEntity();
     if ($entity->isNew() || $entity->getEntityTypeId() != 'media') {
       return $elements;
     }
 
-    $node = $this->utils->getParentNode($entity);
-
-    if ($node === NULL) {
-      return $elements;
+    if ($alt_text_setting == 'none') {
+      foreach ($elements as $element) {
+        $element['#item']->set('alt', '');
+      }
     }
 
-    $url = $node->toUrl();
+    if ($image_link_setting == 'content' || $alt_text_setting == 'original_file' || $alt_text_setting == 'original_file_fallback') {
+      $node = $this->utils->getParentNode($entity);
+      if ($node === NULL) {
+        return $elements;
+      }
 
-    foreach ($elements as &$element) {
-      $element['#url'] = $url;
-    }
+      if ($image_link_setting == 'content') {
+        // Set image link.
+        $url = $node->toUrl();
+        foreach ($elements as &$element) {
+          $element['#url'] = $url;
+        }
+      }
+
+      if ($alt_text_setting == 'original_file' || $alt_text_setting == 'original_file_fallback') {
+        do {
+          $original_file_term = $this->utils->getTermForUri("http://pcdm.org/use#OriginalFile");
 
+          if ($original_file_term === NULL) {
+            break;
+          }
+
+          $original_file_media = $this->utils->getMediaWithTerm($node, $original_file_term);
+
+          if ($original_file_media === NULL) {
+            break;
+          }
+          $source_field_name = $this->mediaSourceService->getSourceFieldName($original_file_media->bundle());
+          if (!$original_file_media->hasField($source_field_name)) {
+            break;
+          }
+          $original_file_files = $original_file_media->get($source_field_name);
+          $alt_text = '';
+          // Get Nth alt text from the Nth file field.
+          $i = 0;
+          foreach ($original_file_files as $file) {
+            if (!$file->hasProperty('alt')) {
```

Comment:
This doesn't exist as a `method`, do we expect that the `ImageItem` might not have `alt` as a property?

Reply:
The original file file might not be an "image" (i.e. a tiff or jp2)